runtime/compiler/optimizer/SequentialStoreSimplifier.cpp

```diff
@@ -677,621 +684,2233 @@ TR::Node* getALoadReferenceForSeqLoad(TR::Node* rootNode, int32_t totalBytes, in
-      case TR::l2b: shift = true;  _valueSize = 8; shiftCode = TR::lushr; altShiftCode = TR::lshr; constCode = TR::iconst; break;
-      case TR::bload:
-      case TR::bconst:
-      case TR::sconst:
-      case TR::iconst:
-      case TR::lconst: shift = false; _valueSize = 1; _shiftValue = 0; break;
-      default: return false;
-      }
-
-//   dumpOptDetails(comp(), "arraycopy sequential store - check load node %p\n", loadNode);
-
-   if (shift)
+   /*
+    * seqLoadSearchAndCombine recursively searches down the tree looking for a set of nodes that potentially matches
+    * the pattern of byte loads to sequential memory locations that are used to construct a larger value.
+    * This search looks at opcodes, const values and node reference counts to check for a match.
+    * If these checks pass, generateArraycopyFromSequentialLoads is called which does a few more detailed check and
+    * potentially performs the transformation to combine the byte loads together.
+    */
+   if (visitedNodes->isSet(currentNode->getGlobalIndex()))
+      return currentTree;
+
+   visitedNodes->set(currentNode->getGlobalIndex());
+
+   /*
+    * Combine nodes are the "add" and "or" nodes that construct the longer value out of several loaded bytes.
+    * combineNodeList keeps tracks of the nodes used to construct one particular value. It is populated by isValidSeqLoadCombine.
+    * combineNodeList needs to be cleared each time before it is repopulated.
+    * isValidSeqLoadCombine also performs a series of initial checks to see if the tree structure contains the pattern
+    * of byte loads that can be combined into wider loads.
+    * combineNodeCount tracks the number of combine nodes.
+    */
+   combineNodeList->clear();
+   int32_t combineNodeCount = 0;
+   if (isValidSeqLoadCombine(comp, trace, currentNode, combineNodeList, combineNodeCount))
       {
-      if (loadNode->getFirstChild()->getOpCodeValue() != shiftCode && loadNode->getFirstChild()->getOpCodeValue() != altShiftCode)
-         {
-         _valNode = loadNode->getFirstChild();
-         _shiftValue = 0;
-         }
-      else
-         {
-         TR::Node* shiftVal = loadNode->getFirstChild()->getSecondChild();
-         if (shiftVal->getOpCodeValue() != constCode)
-            {
-            dumpOptDetails(comp(), " Shifted Value: did not encounter correct const code\n");
-            return false;
-            }
-         _valNode = loadNode->getFirstChild()->getFirstChild();
-         switch(constCode)
-            {
-            case TR::sconst: _shiftValue = shiftVal->getShortInt(); break;
-            case TR::iconst: _shiftValue = shiftVal->getInt(); break;
-            case TR::lconst: _shiftValue = shiftVal->getLongInt(); break;
-            default: break;
-            }
-         }
+      /*
+       * The initial tree structure seems good at this point. generateArraycopyFromSequentialLoads will do some additional
```

Comment:
I decided to just remove this comment since it keeps causing problems. There's already a comment a few lines earlier that mentions what `isValidSeqLoadCombine` and `generateArraycopyFromSequentialLoads` do. Those functions themselves also have large comments that provide the details on what they are doing.